reconcile/gitlab_permissions.py

```diff
@@ -1,21 +1,137 @@
+            project.web_url: self.extract_group_spec(project)
+            for project in projects_from_group
+        }
+        self.reconcile(desired_state, current_state)
+
+    def extract_group_spec(self, project: GroupProject) -> GroupSpec:
+        # check if the project have shared_with_groups with an access level
+        # in case of group being the only ancestor of project the shared_with_groups will be empty
+        for group in project.shared_with_groups:
+            if group["group_id"] == self.group.id:
+                return GroupSpec(
+                    group_name=self.group.name,
+                    group_access_level=group["group_access_level"],
+                )
+        # return the desired groupspec data to skip project whose ancestor is the group as they cannot be shared/unshared
+        return GroupSpec(
+            group_name=self.group.name,
+            group_access_level=self.access_level,
+            is_ancestor=True,
+        )
+
+    def can_share_project(self, project: Project) -> bool:
+        # check if user have access greater or equal access to be shared with the group
+        user = project.members_all.get(id=self.gl.user.id)
+        return user.access_level >= self.access_level
+
+    def reconcile(
+        self,
+        desired_state: dict[str, GroupSpec],
+        current_state: dict[str, GroupSpec],
+    ) -> None:
+        # get the diff data
+        diff_data = diff_mappings(
+            current=current_state,
+            desired=desired_state,
+            equal=lambda current, desired: current.group_access_level
+            == desired.group_access_level,
+        )
+        errors: list[Exception] = []
+        for repo in diff_data.add:
+            project = self.gl.get_project(repo)
+            if not self.can_share_project(project):
+                errors.append(
+                    GroupAccessLevelError(
+                        f"{repo} is not shared with {self.gl.user.username} as {self.access_level_string}"
+                    )
+                )
+                continue
+            logging.info([
+                f"share_group_{self.group.name}_as_{self.access_level_string}",
+                repo,
+            ])
+            if not self.dry_run:
+                self.gl.share_project_with_group(
+                    project=project,
+                    group_id=self.group.id,
+                    access_level=self.access_level,
+                )
+        for repo in diff_data.change:
+            project = self.gl.get_project(repo)
+            if not self.can_share_project(project):
+                errors.append(
+                    GroupAccessLevelError(
+                        f"{repo} is not shared with {self.gl.user.username} as {self.access_level_string}"
+                    )
+                )
+                continue
+            logging.info([
+                f"reshare_group_{self.group.name}_as_{self.access_level_string}",
```

Comment:
too many `_` in logging message

Reply:
same in `reshare_with_`

Reply:
I mean other log messages are not using `_` , any reason for this one to be different?

Reply:
@hemslo now I understand..let me make these logs similar to other integration.

Reply:
```
['reshare', '<repo>', '<group_name>', '<access_level>']
```